tutorial/20_recipes/013_wilcoxon_pruner.py

```diff
@@ -0,0 +1,324 @@
+# the solution far away from the current best solution, and rolling back to the best solution
+# periodically often improves optimization efficiency a lot. However, if the rollback happens
+# too often, the optimization may get stuck in a local optimum, so we must tune the threshold
+# to a sensible amount.
+#
+# .. note::
+#     Some samplers, including the default ``TPESampler``, currently cannot utilize the
+#     information of pruned trials effectively (especially when the last intermediate value
+#     is not the best approximation to the final objective function).
+#     As a workaround for this issue, you can return an estimation of the final value
+#     (e.g., the average of all evaluated values) when ``trial.should_prune()`` returns ``True``,
+#     instead of `raise optuna.TrialPruned()`.
+#     This will improve the sampler performance.
+
+
+def objective(trial: optuna.Trial) -> float:
+    global num_evaluation
+    options = SAOptions(
+        max_iter=N_SA_ITER,
+        T0=trial.suggest_float("T0", 0.01, 10.0, log=True),
+        alpha=trial.suggest_float("alpha", 1.0, 10.0, log=True),
+        patience=trial.suggest_int("patience", 10, 1000, log=True),
+    )
+    results = []
+
+    # For best results, shuffle the evaluation order in each trial.
+    ordering = np.random.permutation(len(dataset))
+    for i in ordering:
+        num_evaluation += 1
+        result_idxs = tsp_simulated_annealing(vertices=dataset[i], options=options)
+        result_cost = tsp_cost(dataset[i], result_idxs)
+        results.append(result_cost)
+
+        trial.report(result_cost, i)
+        if trial.should_prune():
+            # raise optuna.TrialPruned()
+
+            # Return the current predicted value when pruned.
+            # This is a workaround for the problem that
+            # current TPE sampler cannot utilize pruned trials effectively.
+            return sum(results) / len(results)
+
+    return sum(results) / len(results)
+
+
+###################################################################################################
+# We use ``TPESampler`` with ``WilcoxonPruner``.
+
+
+np.random.seed(0)
+sampler = optuna.samplers.TPESampler(seed=1)
+pruner = optuna.pruners.WilcoxonPruner(p_threshold=0.1)
+study = optuna.create_study(direction="minimize", sampler=sampler, pruner=pruner)
+study.enqueue_trial({"T0": 1.0, "alpha": 2.0, "patience": 50})  # default params
+study.optimize(objective, n_trials=N_TRIALS)
+
+
+###################################################################################################
+# We can show the optimization results as:
+
+
+print(f"The number of trials: {len(study.trials)}")
+print(f"Best value: {study.best_value} (params: {study.best_params})")
+print(f"Number of evaluation: {num_evaluation} / {len(dataset) * N_TRIALS}")
+
+
+###################################################################################################
+# Visualize the optimization history.
```

Comment:
It would be helpful to note that this history plot shows both completed and pruned trials in same ways.